File: Test/AdventureWorksFunctionalModel/Sales/CreditCard.cs







using System;
using System.Collections.Generic;
using AW.Functions;
using NakedFunctions;

namespace AW.Types {
    public class CreditCard {
        [Hidden]
        public int CreditCardID { get; init; }

        [Hidden]
        public string CardType { get; init; } = "";

        [Hidden]
        public string CardNumber { get; init; } = "";

        [Hidden]
        public byte ExpMonth { get; init; }

        [Hidden]
        public short ExpYear { get; init; }

        [Named("Persons")] [MemberOrder(5)] [TableView(false, nameof(PersonCreditCard.Person))]
        public virtual ICollection<PersonCreditCard> PersonLinks { get; init; } = new List<PersonCreditCard>();

        [MemberOrder(99)]
        [Versioned]
        public DateTime ModifiedDate { get; init; }

        
        public override string? ToString() => CreditCard_Functions.ObfuscatedNumber(this);


    }
}

File: Test/AdventureWorksLegacyModelVB/Sales/CreditCard.vb


Namespace AW.Types

	Partial Public Class CreditCard

		Implements ITitledObject
		''<Hidden>
		Public Property CreditCardID() As Integer

		''<Hidden>
		Public Property CardType() As String = ""

		''<Hidden>
		Public Property CardNumber() As String = ""

		''<Hidden>
		Public Property ExpMonth() As Byte

		''<Hidden>
		Public Property ExpYear() As Short

#Region "PersonLinks (Collection)"
		Public Overridable Property mappedPersonLinks As ICollection(Of PersonCreditCard) = New List(Of PersonCreditCard)()

		Private myPersonLinks As InternalCollection

		'<TableView(False, NameOf(PersonCreditCard.Person))>
		<MemberOrder(5)>
		Public ReadOnly Property PersonLinks As InternalCollection
			Get
				myPersonLinks = If(myPersonLinks, New InternalCollection(Of PersonCreditCard)(mappedPersonLinks))
				Return myPersonLinks
			End Get
		End Property

		Public Sub AboutPersonLinks(a As IFieldAbout)
			Select Case a.TypeCode
				Case AboutTypeCodes.Name
					a.Name = "Persons"
				Case Else
			End Select
		End Sub
#End Region

#Region "ModifiedDate"
		Public Property mappedModifiedDate As Date
		Friend myModifiedDate As TimeStamp

		<MemberOrder(99)>
		Public ReadOnly Property ModifiedDate As TimeStamp
			Get
				myModifiedDate = If(myModifiedDate, New TimeStamp(mappedModifiedDate, Sub(v) mappedModifiedDate = v))
				Return myModifiedDate
			End Get
		End Property

		Public Sub AboutModifiedDate(a As IFieldAbout)
			Select Case a.TypeCode
				Case AboutTypeCodes.Usable
					a.Usable = False
			End Select
		End Sub
#End Region


		Public Function Title() As Title Implements ITitledObject.Title
			Return New Title(ToString())
		End Function

		Public Overrides Function ToString() As String
			Return If(CardNumber.Length > 4,
				CardNumber.Substring(CardNumber.Length - 4).PadLeft(CardNumber.Length, "*"c),
				Nothing)
		End Function

	End Class
End Namespace